lib/pbench/test/unit/server/test_requests.py

```diff
@@ -127,174 +134,165 @@ def test_malformed_authorization_header(self, client, caplog, server_config):
         assert response.status_code == HTTPStatus.UNAUTHORIZED
         self.verify_logs(caplog)
 
-    def test_missing_controller_header_upload(self, client, caplog, server_config):
-        with client:
-            auth_token = get_pbench_token(client, server_config)
-
-            expected_message = "Missing required controller header"
-            response = client.put(
-                self.gen_uri(server_config),
-                headers={"Authorization": "Bearer " + auth_token},
-            )
-            assert response.status_code == HTTPStatus.BAD_REQUEST
-            assert response.json.get("message") == expected_message
-            self.verify_logs(caplog)
+    def test_missing_controller_header_upload(
+        self, client, caplog, server_config, pbench_token
+    ):
+        expected_message = "Missing required controller header"
+        response = client.put(
+            self.gen_uri(server_config),
+            headers={"Authorization": "Bearer " + pbench_token},
+        )
+        # assert response.json.get("message") == "test"
+        assert response.status_code == HTTPStatus.BAD_REQUEST
+        assert response.json.get("message") == expected_message
+        self.verify_logs(caplog)
 
     def test_missing_md5sum_header_upload(
-        self, client, caplog, server_config, setup_ctrl
+        self, client, caplog, server_config, setup_ctrl, pbench_token
     ):
-        with client:
-            auth_token = get_pbench_token(client, server_config)
-
-            expected_message = "Missing required Content-MD5 header"
-            response = client.put(
-                self.gen_uri(server_config), headers=self.gen_headers(auth_token)
-            )
-            assert response.status_code == HTTPStatus.BAD_REQUEST
-            assert response.json.get("message") == expected_message
-            self.verify_logs(caplog)
+        expected_message = "Missing required Content-MD5 header"
+        response = client.put(
+            self.gen_uri(server_config), headers=self.gen_headers(pbench_token),
+        )
+        assert response.status_code == HTTPStatus.BAD_REQUEST
+        assert response.json.get("message") == expected_message
+        self.verify_logs(caplog)
```

Comment:
This is suspicious:  we need a token in order to test some bad requests but not others?  :-(

Reply:
What you mean?

Reply:
In `test_missing_controller_header_upload()`, we specify an `"Authorization"` header in the PUT request; in `test_missing_md5sum_header_upload()`, we don't.

I understand that we don't want to test every combination of present and absent headers, but it would make the testing of missing headers more comprehensible if every test presented all the headers except for the one whose absence it is testing.

To the greatest extent possible, unit testing should not be implemented based on "whitebox knowledge" of the CUT.  The unit tests should control the inputs to the function or unit being tested, including behavior driven by its dependencies (i.e., by mocking subroutine calls), and they should avoid relying on implementation details within the function (like the order in which headers are checked).  That sort of reliance results in fragile test scenarios which have to be modified when the CUT is modified, which undermines the robustness of the testing -- we want the freedom to modify the CUT without modifying the tests, so that the tests will prove that the new code is still correct.  It's true that unit testing does require significant knowledge of the implementation -- this is required in order to supply appropriate mocks and to ensure that all paths through the CUT are exercised -- but we need to approach this in such a way as to avoid too close a marriage between the tests and the CUT.

Reply:
Actually we are using `gen_headers` utility method to generate headers in `test_missing_md5sum_header_upload` it only takes the `pbench_token` as a parameter and adds it to the set of header values (controller and authorization). We can't use that method in `test_missing_controller_header_upload` because it by defaults add the controller to the header list.

Reply:
I think it would be better if the code did not use `gen_headers()` for the test scenarios which are omitting the various headers.  The function is handy for the other cases, where the headers are more or less a given, but, for the cases where we are explicitly testing what happens when one of them is missing, I recommend that you just construct the header object explicitly -- it's nearly trivial to do (we could almost get by without using `gen_headers()` in the other functions...), and doing it that way makes it obvious to the reader what is and isn't being specified.  And then you can further simplify `gen_headers()` by getting rid of the default value for the hash argument and the conditional code, because it will always be constructing all of the headers.

Reply:
Yeah that make sense